File: week-4/problem-set/prob4.py
"""Prob4 example.

This file belongs to the week 4 divide-and-conquer and memoization examples. It
computes powers of two with a simple recursive definition.
"""


def power_of_two(n):
    """Return 2 raised to the power n using recursion.

    Args:
        n (int): The n input used by this function.

    Returns:
        int: The integer result produced by the algorithm.
    """
    if n == 0:
        return 1
    return 2 * power_of_two(n - 1)


def main():
    """Run the small demonstration for this teaching example.

    Returns:
        None: This function updates data or prints results and does not return a value.
    """
    print(power_of_two(5))


if __name__ == "__main__":
    main()

File: week-4/problem-set/prob4.cpp
/**
 * @file prob4.cpp
 * @brief C++ version of the recursive power-of-two example.
 */

#include <iostream>

using namespace std;

/**
 * @brief Return 2 raised to the power n using recursion.
 * @param n int Exponent of 2.
 * @return int Value of 2^n.
 */
int power_of_two(int n) {
    if (n == 0) {
        return 1;
    }
    return 2 * power_of_two(n - 1);
}

/**
 * @brief Run the power-of-two demonstration.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    cout << power_of_two(5) << '\n';
    return 0;
}
